src/replication.c

```diff
@@ -1649,19 +1669,21 @@ void replicationQuorum(struct raft *r, const raft_index index)
     // assert(logTermOf(r->log, index) > 0);
     assert(logTermOf(r->log, index) <= r->current_term);
 
-    for (i = 0; i < r->configuration.n; i++) {
-        struct raft_server *server = &r->configuration.servers[i];
-        if (server->role != RAFT_VOTER) {
-            continue;
-        }
-        if (r->leader_state.progress[i].match_index >= index) {
-            votes++;
+    /* Raft Paper fig. 2 - Rules for Servers
+     * If there exists an N such that N > commitIndex, a majority
+     * of matchIndex[i] >= N, and log[N].term == currentTerm:
+     * set commitIndex = N (5.3, 5.4).
+     * */
+    for (n = index; n > r->commit_index; n -= 1) {
```

Comment:
This implementation might be a bit naive, if n is large and commit index is 0 e.g. and a majority of the followers their match index hasn't been updated or is low, this loop can run for a long while, I'll first fix this.